File: base/src/main/java/com/ssz/baselibrary/app/helper/AppHelper.java
package com.ssz.baselibrary.app.helper;

import android.app.ActivityManager;
import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.support.annotation.NonNull;

import com.ssz.baselibrary.BuildConfig;
import com.ssz.baselibrary.app.AppContext;
import com.ssz.baselibrary.utils.ObjectHelper;
import com.ssz.baselibrary.utils.ScreenUtil;
import com.ssz.baselibrary.utils.SystemUtil;

import java.util.List;

/**
 * @author zsp
 * create at 2019/12/18 15:07
 * AppHelper
 */
public class AppHelper {

    private static final AppHelper instance = new AppHelper();
    private static Application application;
//    public static final boolean DEBUG = Boolean.parseBoolean("true");
    public static final boolean DEBUG = BuildConfig.DEBUG;

    @AppStatus
    private String appStatus = AppStatus.KILLED;

    private AppHelper(){}

    public static AppHelper init(@NonNull Application application){
        AppHelper.application = application;
        return instance;
    }

    public static AppHelper get(){
        return instance;
    }

    public static Application getApplication(){
        return ObjectHelper.requireNotNull(application,"AppHolder not must be init() first !");
    }

    public static AppContext getAppContext(){
        return ((AppContext) getApplication());
    }

    public void setAppStatus(@AppStatus String status){
        this.appStatus = status;
    }

    @AppStatus
    public String getAppStatus(){
        return appStatus;
    }

    public boolean isKilled(){
        return appStatus.equals(AppStatus.KILLED);
    }

    public AppHelper toLog(){
        if (DEBUG){
            SystemUtil.log();
            ScreenUtil.log();
        }
        return this;
    }

    public static void exitApp(){
        boolean bellowOsFive = Boolean.parseBoolean("true");
        if (bellowOsFive){
            getApplication().sendBroadcast(new Intent(ExitAppBroadcast.EDIT_ACTION));
        }else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP){
                ActivityManager activityManager = (ActivityManager)getApplication().getSystemService(Context.ACTIVITY_SERVICE);
                List<ActivityManager.AppTask> appTaskList = activityManager.getAppTasks();
                for (ActivityManager.AppTask appTask : appTaskList) {
                    appTask.finishAndRemoveTask();
                }
            }
        }
    }

    public void restartApp(){
        final Application application = getApplication();
        Intent intent = application.getPackageManager().getLaunchIntentForPackage(application.getPackageName());
        if (intent != null) {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            application.startActivity(intent);
        }
    }
}

File: app/src/main/java/com/ssz/studydemo/base/app/helper/AppHelper.kt
package com.ssz.studydemo.base.app.helper

import android.app.Application
import android.content.Intent
import com.ssz.studydemo.base.app.helper.ExitAppBroadcast.Companion.EDIT_ACTION
import com.ssz.studydemo.utils.ScreenUtil
import com.ssz.studydemo.utils.SystemUtil

/**
 * @author : zsp
 * time : 2019 12 2019/12/17 9:41
 * 静态持有 Application 类对象，然后全局再使用
 */
object AppHelper {

    /**
     * debug 模式
     */
    const val isDebug = true

    /**
     * 静态持有Application instance
     */
    private var mContext : Application? = null
    fun init(context : Application):AppHelper{
        mContext = context
        return this
    }

    fun getApplication() = mContext ?: throw NullPointerException("AppHolder not must be init() first !")
//    fun getAppContext() = getApplication() as AppContext

    @AppStatus
    private var appStatus :String = AppStatus.KILLED
    fun setAppStatus(@AppStatus status : String){
        this.appStatus = status
    }
    fun getAppStatus() = appStatus
    fun isKilled() = appStatus == AppStatus.KILLED

    fun toLog():AppHelper{
        SystemUtil.log()
        ScreenUtil.log()
        return this
    }

    fun exitApp(){
        getApplication().sendBroadcast(Intent(EDIT_ACTION))

//        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
//            val activityManager = getApplication().getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
//            val appTaskList = activityManager.appTasks
//            for (appTask in appTaskList) {
//                appTask.finishAndRemoveTask()
//            }
//        }
    }

    fun restartApp(){
        val application = getApplication()
        val intent = application.packageManager
                .getLaunchIntentForPackage(application.packageName)
        if (intent != null) {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
            application.startActivity(intent)
        }
    }
}